### ISSUE TITLE
Possible mistake in AGNNConv

### ISSUE THREAD

**Issue by sw-gong** [rusty1s/pytorch_geometric#227]
See the below source code of the class `AGNN`:
```
def message(self, edge_index_i, x_j, x_norm_i, x_norm_j, num_nodes):
        # Compute attention coefficients.
        beta = self.beta if self.requires_grad else self._buffers['beta']
        alpha = beta * (x_norm_i * x_norm_j).sum(dim=-1)
        alpha = softmax(alpha, edge_index_i, num_nodes)

        return x_j * alpha.view(-1, 1)
```

Shouldn't it be `alpha = softmax(alpha, edge_index_j, num_nodes)`? I consider the denominator in terms of the softmax should be the sum of all neighboring nodes connecting to the target node `i` thus `edge_index_j`.

**Comment by rusty1s** [comment]
This is correct IMO. We refer to `_j` as the nodes *to* aggregate and to `_i` as the nodes *which* aggregate (so that it is consistent when changing the message passing flow). This is the reason the `softmax` groups all attention scores across the central nodes which aggregate.

**Comment by sw-gong** [comment]
I'm not quite sure with what you said. In my opinion, nodes `_i` will be aggregated in your `update` stage and `softmax` should give the score of node `_i` with respect to each neighboring node `_j` so that it should group all the neighboring nodes with `edge_index_j`?

**Comment by rusty1s** [comment]
It is actually the other way round. `softmax` should normalize the scores of neighboring nodes `_j` with respect to each central node `_i`.

**Comment by sw-gong** [comment]
Assume the unnormalized attention score matrix `e`, the attention score alpha_ij = exp(e_ij) / sum_{j \in \mathcal(N)_i} exp(e_ij), and adjacency matrix `A`. We can get the attention score with the below code, right? So, `softmax` aggregates the scores of the neighboring nodes thus `edge_index_j` in your case? Or I misunderstood your source code?

```
zero_vec = torch.zeros_like(e)
alpha = torch.where(A > 0, e, zero_vec)
alpha = F.softmax(alpha, dim=1)
```

**Comment by rusty1s** [comment]
That is really a little confusing, but the `idx` of the `softmax` function denotes which nodes are "grouped" together to compute the softmax. Given default `source_to_target` flow, this has to be `edge_index[1]`, hence `edge_index_i`, because we need to group the nodes which share the same target.

**Comment by sw-gong** [comment]
I see. I misunderstood the method `softmax` which aggregates/processes values with the same index.

Sorry. Please close this issue.

**Comment by rusty1s** [comment]
I will try to make this more clear in the doc. Thank you!
